README.md
__GeoIP binding for nodejs__

[![NPM version](https://badge.fury.io/js/geoip.png)](http://badge.fury.io/js/geoip)
[![Build status](https://secure.travis-ci.org/kuno/GeoIP.png)](http://travis-ci.org/kuno/GeoIP)
[![Dependencies Status](https://david-dm.org/kuno/GeoIP.png)](https://david-dm.org/kuno/GeoIP)
[![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/kuno/geoip/trend.png)](https://bitdeli.com/free "Bitdeli Badge")

Get geolocation information based on domain or IP address.

**Pure javascript implementation**

If you are looking for a pure javascript implementation, the [`node-geoip`](https://github.com/bluesmoon/node-geoip) maybe your choice.

###Technical Architecture

![new_architecture](https://github.com/kuno/GeoIP/raw/master/misc/new_architecture.png)

###Memory leak issue
As up to the version 0.4.7 (commit [22dadd98646f75318412b54b3de90cc079d34b1d](https://github.com/kuno/GeoIP/commit/22dadd98646f75318412b54b3de90cc079d34b1d)), there will be some memory leak when new a object from GeoIP class.

Here is the [report](https://gist.github.com/4357339), I'll update it frequently.

Fortunately, the number of leaked bytes is not huge, and you wont new a object very offen.

If you are willing to fix this but, please fork this project.

###Need support special database?
Corrently, this package supports the most common databases from MaxMind.

But, it still a lot of special databases out there.

Since I am not a user of any product/service of maxmind, if you require this pacakge to support special db, please make sure you can delivery the testing db and the ip address/domain that can be used to against the db.


##Installation

### for nodejs >= 0.8.x

    [sudo] npm install geoip [-g]

### for nodejs 0.6.x

    [sudo] npm install geoip@0.4.5 [-g]

### for nodejs 0.4.x

    [sudo] npm install geoip@0.4.4 [-g]

### for nodejs older than 0.4.0

From v0.4.0, geoip need nodejs >= 0.4.0, if you want to use it on old nodejs, you can:

    npm install geoip@0.3.4-1


##Usage

###Check edition

    var edition = geoip.check('/path/to/file');

    console.log(edition); // output 'country', 'city', 'org'... so on

###Create a new instance of sub-module, for example:

    var city = new geoip.City('/path/to/GeoLiteCity.dat', false);  // not to cache the database

    var city = new geoip.City('/path/to/GeoLiteCity.dat');  // the default option is cache

###Update database on the fly:

    city.update('/path/to/new/GeoLiteCity.dat');

##Modules

###Country

ipv4 address lookup

    // Open the country data file
    var Country = geoip.Country;

    var country = new Country('/path/to/GeoIP.dat');

    // Synchronous method(the recommended way):
    var country_obj = country.lookupSync('8.8.8.8');

    console.log(country_obj);
    /*
      { country_code: 'US',
        country_code3: 'USA',
        country_name: 'United States',
        continent_code: 'NA'
      }
    */

    // Asynchronous method:
    country.lookup('www.google.com', function(err, data) {
        if (err) {throw err;}
        if (data) { // if not found, just return null
            console.log(data);  // same as lookup method
        }
    });

ipv6 address lookup (Currently only Country and City module supports ipv6)

    // Open the ipv6 edition of country module
    var Country6 = geoip.Country6;

    var country_v6 = new Country6('/path/to/GeoIPv6.dat');

    // Synchronous method
    var country_obj_v6 = country_v6.lookupSync('2607:f0d0:1002:0051:0000:0000:0000:0004');

    console.log(country_obj_v6); // Same as ipv4
    /*
      { country_code: 'US',
        country_code3: 'USA',
        country_name: 'United States',
        continent_code: 'NA'
      }
    */

    // Asynchronous method
    country_v6.lookup('2400:2352:b0f1:36c5:aa9d:694a:2f98:40bd', function(err, data_v6) {
        if (err) {throw err;}
        if (data_v6) {
            console.log(data_v6);
            /*
             {
               country_code: 'JP',
               country_code3: 'JPN',
               country_name: 'Japan',
               continent_code: 'AS'
              }

            */
        }
    });


###City

ipv4 address lookup

    // Open the GeoLiteCity.dat file first.
    var City = geoip.City;
    var city = new City('/path/to/GeoLiteCity.dat');

  Synchronous method:

      var city_obj = city.lookupSync('8.8.8.8');
      console.log(city_obj);
      /*
      Return an object of city information
      {
       "country_code":"US",
       "country_code3":"USA",
       "country_name":"United States",
       "continet_code":"NA",
       "region":"CA",
       "city":"Mountain View",
       "postal_code":"94043",
       "latitude":37.41919999999999,
       "longitude":-122.0574,
       "dma_code":807,
       "metro_code":807,
       "area_code":650
       }
       */

  Asynchronous method:

      city.lookup('www.google.com', function(err, data) {
        if (err) {throw err;}
        if (data) {
            console.log(data);
        }
      });


ipv6 address lookup

    // Open the GeoLiteCityv6.dat file first.
    var City6 = geoip.City6;
    var city6 = new City6('/path/to/GeoLiteCityv6.dat');

  Synchronous method:

      var city6_obj = city6.lookupSync('2002:7679:b9af:db9b:ab51:501a:db4e:2d');
      console.log(city6_obj);
      /*
      {
       country_code: 'CN',
       country_code3: 'CHN',
       country_name: 'China',
       region: '32',
       city: 'Chengdu',
       latitude: 30.66670036315918,
       longitude: 104.06670379638672,
       continent_code: 'AS'
      }
      */

  Asynchronous method:

      city6.lookup('2002:c6aa:2dc8:b719:4314:2a7d:e31c:fe73', function(err, data) {
        if (err) {throw err;}
        if (data) {
            console.log(data);
            /* As expected ,return an object that contains city information
            {
             country_code: 'US',
             country_code3: 'USA',
             country_name: 'United States',
             region: 'CO',
             city: 'Englewood',
             postal_code: '80112',
             latitude: 39.569000244140625,
             longitude: -104.85820007324219,
             continent_code: 'NA'
            }
            */
          }
      });

###Organization

    var Org = geoip.Org;
    var org = new Org('/path/to/file')  // Org module can open three edition database 'org', 'asnum', 'isp'

Synchronous method:

    var org_str = org.lookupSync('8.8.8.8');

    console.log(org_str);
    /*
      The type of returned data is string, for example:

      'Genuity'
      'AS15169 Google Inc.'

      no longer an object
    */

Asynchronous method:

    org.lookup('www.google.com', function(err, data) {
        if (err) {throw err;}
        if (data) {
            console.log(data);
        }
    });


###Region

    var Region = geoip.Region;
    var region = new Region('/path/to/GeoIPRegion.dat');

Synchronous method:

    var region_obj = region.lookupSync('8.8.8.8');

    console.log('region_obj);
    /*
      region object has two properties:
      { country_code: 'US', region: 'CO' }

    */

Asynchronous method:

    region.lookup('www.google.com', function(err, data) {
        if (err) {throw err;}
        if (data) {
          console.log(data);
        }
    });


###NetSpeed

    var NetSpeed = geoip.NetSpeed;
    var netspeed = new NetSpeed('/path/to/GeoIPNetSpeed.dat');

Synchronous method:

    var netspeed_str = netspeed.lookupSync('8.8.8.8');

    console.log(netspeed_str);
    /*
      netspeed_str just a simple string, 'Dialup', 'Corprate'... so on
    */

Asynchronous method:

    netspeed.lookup('www.google.com', function(err, data) {
        if (err) {throw err;}
        if (data) {
          console.log(data);  // Maybe return 'unknow' or different from lookup method
        }
    });


Changelog
v0.1.0, 2010-11-25
1, Initial release.

v0.1.1, 2010-11-27:
1, Removed GeoIP.dat, due to licence(or money) problem.
2, renamed city property to City, renamed country property to Country.
3, Allow close method to set all properties of an exsiting obj to undefined.

v0.1.2, 2010-12-02:
1, Fixed missing of isIP method in node 0.2.x series.
2, Hidden some unsafe an useless funcions.

v0.1.3, 2010-12-02:
1. Fixed that open and close methos dosen't works in real script.

v0.2.1, 2010-12-03:
1, added support for GeoIPASNum.dat binary data.
2, make org_by_addr method return an array of org name string.

v0.3.0, 2010-12-15:
1, All module now has asynchronous version method.
2, When data is not match with module, throw an error object.
3, When not found  desired data, return an error string instead an error objerct.

v0.3.1, 2010-12-16:
1, Limit the length of the floating part of latitude and langitude to 6 digits.

v0.3.2, 2011-01-21:
1, add geoip.check(data) method.
2, validate country binary data based on 8-types fingerprint.
3, support nodejs v0.3.6.
4, make geoip.close(data) method sync style.

v0.3.3, 2011-02-06:
1, drop support for nodejs 0.2.1, pass testing up to nodejs 0.3.8.
2, add geoip.filter method.
3, re-writed testing shell script.
4, If not found matched data, alway return null.
5, use index.js as default main module.
6, added github bug reporting address.

v0.3.3-1, 2011-02-12
1, make compatibility with nodejs v0.4.0 stable branch.

v0.3.4, 2011-02-15
1, new seekCountry algorithm, speedup to about 3x faster (Vladimir Dronnikov).
2, added simple usage property.
3, improve org_by_domain method output.

v0.3.4-1, 2011-03-07
1, fixed cotinent typo.

V0.4.0beta1, 2011-05-11
switch to geoip C library binding.

v0.4.0beta2, 2011-05-15
1, added error checking in async methods.
2, fixed latitude and longitude range problems.
3, added city property in record object.

v0.4.0rc, 2011-05-20
1, implemented basic ipv6 support in Country module.
2, fixed vavious typos.

v0.4.0, 2011-05-23
1, added country ipv6 samples.
2, clear some redundant codes.

v0.4.0final, 2011-05-24
1, change version to 0.4.0final.

v0.4.1, 2011-06-21
1, build addon during installation not preinstall.

v0.4.2, 2011-07-23
1, enbale update database on the fly.
2, disable close method.
3, removed all sleep function.
4, always execute lookup in the main thread.
5, separate county ipv6 into a indiviual module.

v0.4.2-1, 2011-07-24
1, added James Taylor as contributor.
2, added libgeoip's version literature.

v0.4.2-2, 2011-08-01
1, disabled libgeoip version literature due to compatibility issue.

v0.4.3, 2011-08-03
1, enabled converting city name to utf8 characters.

v0.4.4, 2011-09-15
1, support lookup city information through ipv6 address.

v0.4.5, 2011-11-07
1, now needs nodejs >= 0.6.0

v0.4.6, 2012-07-29
1, fully support nodejs 0.8.x, breaking the compatibility with 0.6.x.

v0.4.7*, 2012-12-22
1, fixed most memory leak
2, some cleanup

v0.4.7, 2013-01-04
1, fix serious sagfault issue, re-release v0.4.7.

v0.4.8, 2013-02-15
1, fix missing city data when looking up some special ip in async mode.

v0.4.9, 2013-03-12
1, make compatible with node 0.10.0.

V0.4.10, 2013-05-04
1, Make package self-building, i.e not depending on external libgeoip any
more.

V0.4.11, 2013-10-29
1, Use nan to solved version dependency.
2, Fix #60, Bug in the async country lookup method.

v0.4.12, 2013-11-14
1, Update geoip c api to v1.5.2.


gi.js
var GEOIP = GEOIP || {};
GEOIP.prototype = Object.prototype;
GEOIP.prototype.flags = undefined;
GEOIP.prototype.filehandle = undefined;
GEOIP.prototype.memory_buffer = undefined;
GEOIP.prototype.databaseType = undefined;
GEOIP.prototype.databaseSegments = undefined;
GEOIP.prototype.record_length = undefined;
GEOIP.prototype.shmid = undefined;
GEOIP.prototype.COUNTRY_CODE_TO_NUMBER = [
{"":0}, {"AP":1}, {"EU":2}, {"AD":3}, {"AE":4}, {"AF":5}, 
{"AG":6}, {"AI":7}, {"AL":8}, {"AM":9}, {"AN":10}, {"AO":11}, 
{"AQ":12}, {"AR":13}, {"AS":14}, {"AT":15}, {"AU":16}, {"AW":17}, 
{"AZ":18}, {"BA":19}, {"BB":20}, {"BD":21}, {"BE":22}, {"BF":23}, 
{"BG":24}, {"BH":25}, {"BI":26}, {"BJ":27}, {"BM":28}, {"BN":29}, 
{"BO":30}, {"BR":31}, {"BS":32}, {"BT":33}, {"BV":34}, {"BW":35}, 
{"BY":36}, {"BZ":37}, {"CA":38}, {"CC":39}, {"CD":40}, {"CF":41}, 
{"CG":42}, {"CH":43}, {"CI":44}, {"CK":45}, {"CL":46}, {"CM":47}, 
{"CN":48}, {"CO":49}, {"CR":50}, {"CU":51}, {"CV":52}, {"CX":53}, 
{"CY":54}, {"CZ":55}, {"DE":56}, {"DJ":57}, {"DK":58}, {"DM":59}, 
{"DO":60}, {"DZ":61}, {"EC":62}, {"EE":63}, {"EG":64}, {"EH":65}, 
{"ER":66}, {"ES":67}, {"ET":68}, {"FI":69}, {"FJ":70}, {"FK":71}, 
{"FM":72}, {"FO":73}, {"FR":74}, {"FX":75}, {"GA":76}, {"GB":77},
{"GD":78}, {"GE":79}, {"GF":80}, {"GH":81}, {"GI":82}, {"GL":83}, 
{"GM":84}, {"GN":85}, {"GP":86}, {"GQ":87}, {"GR":88}, {"GS":89}, 
{"GT":90}, {"GU":91}, {"GW":92}, {"GY":93}, {"HK":94}, {"HM":95}, 
{"HN":96}, {"HR":97}, {"HT":98}, {"HU":99}, {"ID":100}, {"IE":101}, 
{"IL":102}, {"IN":103}, {"IO":104}, {"IQ":105}, {"IR":106}, {"IS":107}, 
{"IT":108}, {"JM":109}, {"JO":110}, {"JP":111}, {"KE":112}, {"KG":113}, 
{"KH":114}, {"KI":115}, {"KM":116}, {"KN":117}, {"KP":118}, {"KR":119}, 
{"KW":120}, {"KY":121}, {"KZ":122}, {"LA":123}, {"LB":124}, {"LC":125}, 
{"LI":126}, {"LK":127}, {"LR":128}, {"LS":129}, {"LT":130}, {"LU":131}, 
{"LV":132}, {"LY":133}, {"MA":134}, {"MC":135}, {"MD":136}, {"MG":137}, 
{"MH":138}, {"MK":139}, {"ML":140}, {"MM":141}, {"MN":142}, {"MO":143}, 
{"MP":144}, {"MQ":145}, {"MR":146}, {"MS":147}, {"MT":148}, {"MU":149}, 
{"MV":150}, {"MW":151}, {"MX":152}, {"MY":153}, {"MZ":154}, {"NA":155},
{"NC":156}, {"NE":157}, {"NF":158}, {"NG":159}, {"NI":160}, {"NL":161}, 
{"NO":162}, {"NP":163}, {"NR":164}, {"NU":165}, {"NZ":166}, {"OM":167}, 
{"PA":168}, {"PE":169}, {"PF":170}, {"PG":171}, {"PH":172}, {"PK":173}, 
{"PL":174}, {"PM":175}, {"PN":176}, {"PR":177}, {"PS":178}, {"PT":179}, 
{"PW":180}, {"PY":181}, {"QA":182}, {"RE":183}, {"RO":184}, {"RU":185}, 
{"RW":186}, {"SA":187}, {"SB":188}, {"SC":189}, {"SD":190}, {"SE":191}, 
{"SG":192}, {"SH":193}, {"SI":194}, {"SJ":195}, {"SK":196}, {"SL":197}, 
{"SM":198}, {"SN":199}, {"SO":200}, {"SR":201}, {"ST":202}, {"SV":203}, 
{"SY":204}, {"SZ":205}, {"TC":206}, {"TD":207}, {"TF":208}, {"TG":209}, 
{"TH":210}, {"TJ":211}, {"TK":212}, {"TM":213}, {"TN":214}, {"TO":215}, 
{"TL":216}, {"TR":217}, {"TT":218}, {"TV":219}, {"TW":220}, {"TZ":221}, 
{"UA":222}, {"UG":223}, {"UM":224}, {"US":225}, {"UY":226}, {"UZ":227}, 
{"VA":228}, {"VC":229}, {"VE":230}, {"VG":231}, {"VI":232}, {"VN":233},
{"VU":234}, {"WF":235}, {"WS":236}, {"YE":237}, {"YT":238}, {"RS":239}, 
{"ZA":240}, {"ZM":241}, {"ME":242}, {"ZW":243}, {"A1":244}, {"A2":245}, 
{"O1":246}, {"AX":247}, {"GG":248}, {"IM":249}, {"JE":250}, {"BL":251},
{"MF":252}
];

GEOIP.prototype.COUNTRY_CODES = [
"", "AP", "EU", "AD", "AE", "AF", "AG", "AI", "AL", "AM", "AN", "AO", "AQ",
    "AR", "AS", "AT", "AU", "AW", "AZ", "BA", "BB", "BD", "BE", "BF", "BG", "BH",
    "BI", "BJ", "BM", "BN", "BO", "BR", "BS", "BT", "BV", "BW", "BY", "BZ", "CA",
    "CC", "CD", "CF", "CG", "CH", "CI", "CK", "CL", "CM", "CN", "CO", "CR", "CU",
    "CV", "CX", "CY", "CZ", "DE", "DJ", "DK", "DM", "DO", "DZ", "EC", "EE", "EG",
    "EH", "ER", "ES", "ET", "FI", "FJ", "FK", "FM", "FO", "FR", "FX", "GA", "GB",
    "GD", "GE", "GF", "GH", "GI", "GL", "GM", "GN", "GP", "GQ", "GR", "GS", "GT",
    "GU", "GW", "GY", "HK", "HM", "HN", "HR", "HT", "HU", "ID", "IE", "IL", "IN",
    "IO", "IQ", "IR", "IS", "IT", "JM", "JO", "JP", "KE", "KG", "KH", "KI", "KM",
    "KN", "KP", "KR", "KW", "KY", "KZ", "LA", "LB", "LC", "LI", "LK", "LR", "LS",
    "LT", "LU", "LV", "LY", "MA", "MC", "MD", "MG", "MH", "MK", "ML", "MM", "MN",
    "MO", "MP", "MQ", "MR", "MS", "MT", "MU", "MV", "MW", "MX", "MY", "MZ", "NA",
    "NC", "NE", "NF", "NG", "NI", "NL", "NO", "NP", "NR", "NU", "NZ", "OM", "PA",
    "PE", "PF", "PG", "PH", "PK", "PL", "PM", "PN", "PR", "PS", "PT", "PW", "PY",
    "QA", "RE", "RO", "RU", "RW", "SA", "SB", "SC", "SD", "SE", "SG", "SH", "SI",
    "SJ", "SK", "SL", "SM", "SN", "SO", "SR", "ST", "SV", "SY", "SZ", "TC", "TD",
    "TF", "TG", "TH", "TJ", "TK", "TM", "TN", "TO", "TL", "TR", "TT", "TV", "TW",
    "TZ", "UA", "UG", "UM", "US", "UY", "UZ", "VA", "VC", "VE", "VG", "VI", "VN",
    "VU", "WF", "WS", "YE", "YT", "RS", "ZA", "ZM", "ME", "ZW", "A1", "A2", "O1",
    "AX", "GG", "IM", "JE", "BL", "MF"
    ];

    GEOIP.prototype.COUNTRY_CODES3 = [
    "","AP","EU","AND","ARE","AFG","ATG","AIA","ALB","ARM","ANT","AGO","ATA","ARG",
    "ASM","AUT","AUS","ABW","AZE","BIH","BRB","BGD","BEL","BFA","BGR","BHR","BDI",
    "BEN","BMU","BRN","BOL","BRA","BHS","BTN","BVT","BWA","BLR","BLZ","CAN","CCK",
    "COD","CAF","COG","CHE","CIV","COK","CHL","CMR","CHN","COL","CRI","CUB","CPV",
    "CXR","CYP","CZE","DEU","DJI","DNK","DMA","DOM","DZA","ECU","EST","EGY","ESH",
    "ERI","ESP","ETH","FIN","FJI","FLK","FSM","FRO","FRA","FX","GAB","GBR","GRD",
    "GEO","GUF","GHA","GIB","GRL","GMB","GIN","GLP","GNQ","GRC","SGS","GTM","GUM",
    "GNB","GUY","HKG","HMD","HND","HRV","HTI","HUN","IDN","IRL","ISR","IND","IOT",
    "IRQ","IRN","ISL","ITA","JAM","JOR","JPN","KEN","KGZ","KHM","KIR","COM","KNA",
    "PRK","KOR","KWT","CYM","KAZ","LAO","LBN","LCA","LIE","LKA","LBR","LSO","LTU",
    "LUX","LVA","LBY","MAR","MCO","MDA","MDG","MHL","MKD","MLI","MMR","MNG","MAC",
    "MNP","MTQ","MRT","MSR","MLT","MUS","MDV","MWI","MEX","MYS","MOZ","NAM","NCL",
    "NER","NFK","NGA","NIC","NLD","NOR","NPL","NRU","NIU","NZL","OMN","PAN","PER",
    "PYF","PNG","PHL","PAK","POL","SPM","PCN","PRI","PSE","PRT","PLW","PRY","QAT",
    "REU","ROU","RUS","RWA","SAU","SLB","SYC","SDN","SWE","SGP","SHN","SVN","SJM",
    "SVK","SLE","SMR","SEN","SOM","SUR","STP","SLV","SYR","SWZ","TCA","TCD","ATF",
    "TGO","THA","TJK","TKL","TLS","TKM","TUN","TON","TUR","TTO","TUV","TWN","TZA",
    "UKR","UGA","UMI","USA","URY","UZB","VAT","VCT","VEN","VGB","VIR","VNM","VUT",
    "WLF","WSM","YEM","MYT","SRB","ZAF","ZMB","MNE","ZWE","A1","A2","O1",
    "ALA","GGY","IMN","JEY","BLM","MAF"  
    ];

    GEOIP.prototype.COUNTRY_NAMES = [
    "", "Asia/Pacific Region", "Europe", "Andorra", "United Arab Emirates",
    "Afghanistan", "Antigua and Barbuda", "Anguilla", "Albania", "Armenia",
    "Netherlands Antilles", "Angola", "Antarctica", "Argentina", "American Samoa",
    "Austria", "Australia", "Aruba", "Azerbaijan", "Bosnia and Herzegovina",
    "Barbados", "Bangladesh", "Belgium", "Burkina Faso", "Bulgaria", "Bahrain",
    "Burundi", "Benin", "Bermuda", "Brunei Darussalam", "Bolivia", "Brazil",
    "Bahamas", "Bhutan", "Bouvet Island", "Botswana", "Belarus", "Belize",
    "Canada", "Cocos (Keeling) Islands", "Congo, The Democratic Republic of the",
    "Central African Republic", "Congo", "Switzerland", "Cote D'Ivoire", "Cook Islands",
    "Chile", "Cameroon", "China", "Colombia", "Costa Rica", "Cuba", "Cape Verde",
    "Christmas Island", "Cyprus", "Czech Republic", "Germany", "Djibouti",
    "Denmark", "Dominica", "Dominican Republic", "Algeria", "Ecuador", "Estonia",
    "Egypt", "Western Sahara", "Eritrea", "Spain", "Ethiopia", "Finland", "Fiji",
    "Falkland Islands (Malvinas)", "Micronesia, Federated States of", "Faroe Islands",
    "France", "France, Metropolitan", "Gabon", "United Kingdom",
    "Grenada", "Georgia", "French Guiana", "Ghana", "Gibraltar", "Greenland",
    "Gambia", "Guinea", "Guadeloupe", "Equatorial Guinea", "Greece", "South Georgia and the South Sandwich Islands",
    "Guatemala", "Guam", "Guinea-Bissau",
    "Guyana", "Hong Kong", "Heard Island and McDonald Islands", "Honduras",
    "Croatia", "Haiti", "Hungary", "Indonesia", "Ireland", "Israel", "India",
    "British Indian Ocean Territory", "Iraq", "Iran, Islamic Republic of",
    "Iceland", "Italy", "Jamaica", "Jordan", "Japan", "Kenya", "Kyrgyzstan",
    "Cambodia", "Kiribati", "Comoros", "Saint Kitts and Nevis", "Korea, Democratic People's Republic of",
    "Korea, Republic of", "Kuwait", "Cayman Islands",
    "Kazakhstan", "Lao People's Democratic Republic", "Lebanon", "Saint Lucia",
    "Liechtenstein", "Sri Lanka", "Liberia", "Lesotho", "Lithuania", "Luxembourg",
    "Latvia", "Libyan Arab Jamahiriya", "Morocco", "Monaco", "Moldova, Republic of",
    "Madagascar", "Marshall Islands", "Macedonia",
    "Mali", "Myanmar", "Mongolia", "Macau", "Northern Mariana Islands",
    "Martinique", "Mauritania", "Montserrat", "Malta", "Mauritius", "Maldives",
    "Malawi", "Mexico", "Malaysia", "Mozambique", "Namibia", "New Caledonia",
    "Niger", "Norfolk Island", "Nigeria", "Nicaragua", "Netherlands", "Norway",
    "Nepal", "Nauru", "Niue", "New Zealand", "Oman", "Panama", "Peru", "French Polynesia",
    "Papua New Guinea", "Philippines", "Pakistan", "Poland", "Saint Pierre and Miquelon",
    "Pitcairn Islands", "Puerto Rico", "Palestinian Territory",
    "Portugal", "Palau", "Paraguay", "Qatar", "Reunion", "Romania",
    "Russian Federation", "Rwanda", "Saudi Arabia", "Solomon Islands",
    "Seychelles", "Sudan", "Sweden", "Singapore", "Saint Helena", "Slovenia",
    "Svalbard and Jan Mayen", "Slovakia", "Sierra Leone", "San Marino", "Senegal",
    "Somalia", "Suriname", "Sao Tome and Principe", "El Salvador", "Syrian Arab Republic",
    "Swaziland", "Turks and Caicos Islands", "Chad", "French Southern Territories",
    "Togo", "Thailand", "Tajikistan", "Tokelau", "Turkmenistan",
    "Tunisia", "Tonga", "Timor-Leste", "Turkey", "Trinidad and Tobago", "Tuvalu",
    "Taiwan", "Tanzania, United Republic of", "Ukraine",
    "Uganda", "United States Minor Outlying Islands", "United States", "Uruguay",
    "Uzbekistan", "Holy See (Vatican City State)", "Saint Vincent and the Grenadines",
    "Venezuela", "Virgin Islands, British", "Virgin Islands, U.S.",
    "Vietnam", "Vanuatu", "Wallis and Futuna", "Samoa", "Yemen", "Mayotte",
    "Serbia", "South Africa", "Zambia", "Montenegro", "Zimbabwe",
    "Anonymous Proxy","Satellite Provider","Other",
    "Aland Islands","Guernsey","Isle of Man","Jersey","Saint Barthelemy","Saint Martin"  
    ];

    GEOIP.prototype.CONTINENT_CODES = [
    "--", "AS", "EU", "EU", "AS", "AS", "NA", "NA", "EU", "AS", 
    "NA", "AF", "AN", "SA", "OC", "EU", "OC", "NA", "AS", "EU", 
    "NA", "AS", "EU", "AF", "EU", "AS", "AF", "AF", "NA", "AS", 
    "SA", "SA", "NA", "AS", "AN", "AF", "EU", "NA", "NA", "AS", 
    "AF", "AF", "AF", "EU", "AF", "OC", "SA", "AF", "AS", "SA", 
    "NA", "NA", "AF", "AS", "AS", "EU", "EU", "AF", "EU", "NA", 
    "NA", "AF", "SA", "EU", "AF", "AF", "AF", "EU", "AF", "EU", 
    "OC", "SA", "OC", "EU", "EU", "EU", "AF", "EU", "NA", "AS", 
    "SA", "AF", "EU", "NA", "AF", "AF", "NA", "AF", "EU", "AN", 
    "NA", "OC", "AF", "SA", "AS", "AN", "NA", "EU", "NA", "EU", 
    "AS", "EU", "AS", "AS", "AS", "AS", "AS", "EU", "EU", "NA", 
    "AS", "AS", "AF", "AS", "AS", "OC", "AF", "NA", "AS", "AS", 
    "AS", "NA", "AS", "AS", "AS", "NA", "EU", "AS", "AF", "AF", 
    "EU", "EU", "EU", "AF", "AF", "EU", "EU", "AF", "OC", "EU", 
    "AF", "AS", "AS", "AS", "OC", "NA", "AF", "NA", "EU", "AF", 
    "AS", "AF", "NA", "AS", "AF", "AF", "OC", "AF", "OC", "AF", 
    "NA", "EU", "EU", "AS", "OC", "OC", "OC", "AS", "NA", "SA", 
    "OC", "OC", "AS", "AS", "EU", "NA", "OC", "NA", "AS", "EU", 
    "OC", "SA", "AS", "AF", "EU", "EU", "AF", "AS", "OC", "AF", 
    "AF", "EU", "AS", "AF", "EU", "EU", "EU", "AF", "EU", "AF", 
    "AF", "SA", "AF", "NA", "AS", "AF", "NA", "AF", "AN", "AF", 
    "AS", "AS", "OC", "AS", "AF", "OC", "AS", "EU", "NA", "OC", 
    "AS", "AF", "EU", "AF", "OC", "NA", "SA", "AS", "EU", "NA", 
    "SA", "NA", "NA", "AS", "OC", "OC", "OC", "AS", "AF", "EU", 
    "AF", "AF", "EU", "AF", "--", "--", "--", "EU", "EU", "EU", 
    "EU", "NA", "NA" 
    ];

    exports.GEOIP = GEOIP;
